File: plugin/al/usdtransaction/AL/usd/transaction/tests/testTransaction.cpp
#include "AL/usd/transaction/Notice.h"
#include "AL/usd/transaction/Transaction.h"

#include <pxr/pxr.h>
#include <pxr/usd/usd/attribute.h>
#include <pxr/usd/usd/stage.h>

#include <gtest/gtest.h>

using namespace AL::usd::transaction;
PXR_NAMESPACE_USING_DIRECTIVE

//----------------------------------------------------------------------------------------------------------------------
/// \brief  Test some of the functionality of the AL_USDTransaction
//----------------------------------------------------------------------------------------------------------------------

/// shorthand function for empty paths vector
SdfPathVector empty() { return SdfPathVector(); }

/// helper function to return sorted paths vector from list of strings
SdfPathVector sorted(const std::initializer_list<const char*>& paths)
{
    SdfPathVector result(paths.size());
    std::transform(
        paths.begin(), paths.end(), result.begin(), [](const char* path) { return SdfPath(path); });
    std::sort(result.begin(), result.end());
    return result;
}

/// helper function to return sorted paths vector
SdfPathVector sorted(const SdfPathVector& paths)
{
    SdfPathVector result = paths;
    std::sort(result.begin(), result.end());
    return result;
}

// The fixture for testing transactions.
class TransactionTest
    : public TfWeakBase
    , public ::testing::Test
{
public:
    TransactionTest() { }
    ~TransactionTest() override { }

    size_t               opened() const { return m_opened; }
    size_t               closed() const { return m_closed; }
    const SdfPathVector& getChanged() const { return m_changed; }
    const SdfPathVector& getResynced() const { return m_resynced; }

    /// helper method to create prim at given path with attribute
    void createPrimWithAttribute(const char* path, const char* attributeName = "prop")
    {
        auto prim = m_stage->DefinePrim(SdfPath(path));
        EXPECT_TRUE(prim);
        if (attributeName) {
            auto attr = prim.CreateAttribute(TfToken(attributeName), SdfValueTypeNames->Int);
            EXPECT_TRUE(attr.Set(1));
        }
    }

    /// helper method to change prim attribute at given path to given value
    void changePrimAttribute(const char* path, int value, const char* attributeName = "prop")
    {
        auto prim = m_stage->GetPrimAtPath(SdfPath(path));
        EXPECT_TRUE(prim);
        auto attr = prim.GetAttribute(TfToken(attributeName));
        EXPECT_TRUE(attr.Set(value));
    };

protected:
    void openNotification(const OpenNotice& notice, const UsdStageWeakPtr& stage)
    {
        EXPECT_EQ(stage, m_stage);
        ++m_opened;
    }

    void closeNotification(const CloseNotice& notice, const UsdStageWeakPtr& stage)
    {
        EXPECT_EQ(stage, m_stage);
        ++m_closed;
        m_changed = notice.GetChangedInfoOnlyPaths();
        m_resynced = notice.GetResyncedPaths();
    }

    void SetUp() override
    {
        m_stage = UsdStage::CreateInMemory();
        m_stage->SetEditTarget(m_stage->GetSessionLayer());
        TfWeakPtr<TransactionTest> self(this);
        m_openNoticeKey = TfNotice::Register(self, &TransactionTest::openNotification, m_stage);
        ASSERT_TRUE(m_openNoticeKey);
        m_closeNoticeKey = TfNotice::Register(self, &TransactionTest::closeNotification, m_stage);
        ASSERT_TRUE(m_closeNoticeKey);
    }

    void TearDown() override
    {
        TfNotice::Revoke(m_openNoticeKey);
        TfNotice::Revoke(m_closeNoticeKey);
    }

    UsdStageRefPtr m_stage;

private:
    TfNotice::Key m_openNoticeKey;
    TfNotice::Key m_closeNoticeKey;
    size_t        m_opened = 0;
    size_t        m_closed = 0;
    SdfPathVector m_changed;
    SdfPathVector m_resynced;
};

/// Test that Transaction Open / Close methods work as expected
TEST_F(TransactionTest, Transaction)
{
    Transaction transaction(m_stage, m_stage->GetSessionLayer());
    EXPECT_EQ(opened(), 0u);
    EXPECT_EQ(closed(), 0u);
    // Open notice should be triggered
    EXPECT_TRUE(transaction.Open());
    EXPECT_EQ(opened(), 1u);
    EXPECT_EQ(closed(), 0u);
    // Opening same transaction is allowed, but should not trigger notices
    EXPECT_TRUE(transaction.Open());
    EXPECT_EQ(opened(), 1u);
    EXPECT_EQ(closed(), 0u);
    // Close notices should not be emitted until last close
    EXPECT_TRUE(transaction.Close());
    EXPECT_EQ(opened(), 1u);
    EXPECT_EQ(closed(), 0u);
    // Close notice should be triggered
    EXPECT_TRUE(transaction.Close());
    EXPECT_EQ(opened(), 1u);
    EXPECT_EQ(closed(), 1u);
    // This should fail and no notices should be sent
    EXPECT_FALSE(transaction.Close());
    EXPECT_EQ(opened(), 1u);
    EXPECT_EQ(closed(), 1u);
}

/// Test that ScopedTransaction works as expected
TEST_F(TransactionTest, ScopedTransaction)
{
    EXPECT_EQ(opened(), 0u);
    {
        ScopedTransaction outer(m_stage, m_stage->GetSessionLayer());
        EXPECT_EQ(opened(), 1u);
        EXPECT_EQ(closed(), 0u);
        {
            /// Opening a transaction for same layer should not trigger notices
            ScopedTransaction inner(m_stage, m_stage->GetSessionLayer());
            EXPECT_EQ(opened(), 1u);
            EXPECT_EQ(closed(), 0u);
        }
        EXPECT_EQ(opened(), 1u);
        EXPECT_EQ(closed(), 0u);
        {
            /// Opening a transaction for different layer should trigger notices
            ScopedTransaction inner(m_stage, m_stage->GetRootLayer());
            EXPECT_EQ(opened(), 2u);
            EXPECT_EQ(closed(), 0u);
        }
        EXPECT_EQ(opened(), 2u);
        EXPECT_EQ(closed(), 1u);
    }
    EXPECT_EQ(closed(), 2u);
}

/// Test that CloseNotice reports changes as expected
TEST_F(TransactionTest, Changes)
{
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/A");
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/A" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        changePrimAttribute("/A", 2);
    }
    EXPECT_EQ(sorted(getChanged()), sorted({ "/A.prop" }));
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        changePrimAttribute("/A", 4);
        changePrimAttribute("/A", 2); /// effectively no change
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
}

/// Test that CloseNotice reports hierarchy changes as expected
TEST_F(TransactionTest, Hierarchy)
{
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root");
        createPrimWithAttribute("/root/A");
        createPrimWithAttribute("/root/A/C");
        createPrimWithAttribute("/root/A/D");
        createPrimWithAttribute("/root/B");
        createPrimWithAttribute("/root/B/E");
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        changePrimAttribute("/root", 2);
        changePrimAttribute("/root/A/D", 2);
    }
    EXPECT_EQ(sorted(getChanged()), sorted({ "/root.prop", "/root/A/D.prop" }));
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        changePrimAttribute("/root/B", 2);
        changePrimAttribute("/root/A/C", 2);
        changePrimAttribute("/root/A/C", 1); /// effectively no change
    }
    EXPECT_EQ(sorted(getChanged()), sorted({ "/root/B.prop" }));
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root/B/F");
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root/B/F" }));
}

/// Test that CloseNotice reports property changes as expected
TEST_F(TransactionTest, Properties)
{
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root", "foo");
        createPrimWithAttribute("/root/A", nullptr);
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root", "bar");
        createPrimWithAttribute("/root/A", "foo");
    }
    EXPECT_EQ(sorted(getChanged()), sorted({ "/root.bar", "/root/A.foo" }));
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        changePrimAttribute("/root", 2, "foo");
        changePrimAttribute("/root", 4, "bar");
        changePrimAttribute("/root", 1, "bar"); /// effectively no change
        changePrimAttribute("/root/A", 2, "foo");
    }
    EXPECT_EQ(sorted(getChanged()), sorted({ "/root.foo", "/root/A.foo" }));
    EXPECT_EQ(sorted(getResynced()), empty());
}

/// Test that CloseNotice reports clearing layers as expected
TEST_F(TransactionTest, Clear)
{
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root");
        createPrimWithAttribute("/root/A");
        createPrimWithAttribute("/root/A/B");
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        m_stage->GetSessionLayer()->Clear();
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        createPrimWithAttribute("/root");
        createPrimWithAttribute("/root/A");
        createPrimWithAttribute("/root/A/B");
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), sorted({ "/root" }));
    {
        ScopedTransaction transaction(m_stage, m_stage->GetSessionLayer());
        m_stage->GetSessionLayer()->Clear();
        createPrimWithAttribute("/root");
        createPrimWithAttribute("/root/A");
        createPrimWithAttribute("/root/A/B");
        /// effectively no change
    }
    EXPECT_EQ(sorted(getChanged()), empty());
    EXPECT_EQ(sorted(getResynced()), empty());
}

File: plugin/al/usdtransaction/AL/usd/transaction/tests/testTransaction.py
from AL.usd import transaction
from pxr import Usd, Sdf, Tf
import sys
import unittest

class TestTransaction(unittest.TestCase):

    def setUp(self):
        self._stage = Usd.Stage.CreateInMemory()
        self._stage.SetEditTarget(self._stage.GetSessionLayer())
        self._openNoticeKey = Tf.Notice.Register(transaction.OpenNotice, self._openNoticeHandler, self._stage)
        self._closeNoticeKey = Tf.Notice.Register(transaction.CloseNotice, self._closeNoticeHandler, self._stage)
        self._opened = 0
        self._closed = 0
        self._changed = []
        self._resynced = []
        # assertCountEqual in python 3 is equivalent to assertItemsEqual
        if sys.version_info[0] >= 3:
            self.assertItemsEqual = self.assertCountEqual

    def tearDown(self):
        transaction.TransactionManager.CloseAll()

    def _openNoticeHandler(self, notice, stage):
        self.assertEqual(stage, self._stage)
        self._opened += 1

    def _closeNoticeHandler(self, notice, stage):
        self.assertEqual(stage, self._stage)
        self._changed = notice.GetChangedInfoOnlyPaths()
        self._resynced = notice.GetResyncedPaths()
        self._closed += 1

    def createPrimWithAttribute(self, path, attrName='prop'):
        prim = self._stage.DefinePrim(path)
        self.assertTrue(prim)
        if attrName:
            attr = prim.CreateAttribute(attrName, Sdf.ValueTypeNames.Int)
            self.assertTrue(attr.Set(1))
    
    def changePrimAttribute(self, path, value, attrName='prop'):
        prim = self._stage.GetPrimAtPath(path)
        self.assertTrue(prim)
        attr = prim.GetAttribute(attrName)
        self.assertTrue(attr.Set(value))

    ## Test that Transaction Open / Close methods work as expected
    def test_Transaction(self):
        t = transaction.Transaction(self._stage, self._stage.GetSessionLayer())
        ## This should fail and no notices should be sent
        self.assertFalse(t.Close())
        self.assertEqual(self._opened, 0)
        self.assertEqual(self._closed, 0)
        ## Open notice should be triggered
        self.assertTrue(t.Open())
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 0)
        ## Opening same transaction is allowed, but should not trigger notices
        self.assertTrue(t.Open())
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 0)
        ## Close notices should not be emitted until last close
        self.assertTrue(t.Close())
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 0)
        ## Close notice should be triggered
        self.assertTrue(t.Close())
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 1)
        ## This should fail and no notices should be sent
        self.assertFalse(t.Close())
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 1)

    ## Test that ScopedTransaction for same stage and layer pair work as expected
    def test_SameScopedTransaction(self):
        self.assertEqual(self._opened, 0)
        self.assertEqual(self._closed, 0)
        ## Open notice should be triggered
        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.assertEqual(self._opened, 1)
            self.assertEqual(self._closed, 0)
            ## Opening same transaction is allowed, but should not trigger notices
            with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
                self.assertEqual(self._opened, 1)
                self.assertEqual(self._closed, 0)
            ## Close notices should not be emitted until last close
            self.assertEqual(self._opened, 1)
            self.assertEqual(self._closed, 0)
        ## Close notice should be triggered
        self.assertEqual(self._opened, 1)
        self.assertEqual(self._closed, 1)

    ## Test that ScopedTransaction for different stage and layer pair work as expected
    def test_DifferentScopedTransaction(self):
        self.assertEqual(self._opened, 0)
        self.assertEqual(self._closed, 0)
        ## Open notice should be triggered
        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.assertEqual(self._opened, 1)
            self.assertEqual(self._closed, 0)
            ## Open notice should be triggered as it targets different layer
            with transaction.ScopedTransaction(self._stage, self._stage.GetRootLayer()):
                self.assertEqual(self._opened, 2)
                self.assertEqual(self._closed, 0)
            ## Close notice should be triggered
            self.assertEqual(self._opened, 2)
            self.assertEqual(self._closed, 1)
        ## Close notice should be triggered
        self.assertEqual(self._opened, 2)
        self.assertEqual(self._closed, 2)
    
    ## Test that CloseNotice reports changes as expected
    def test_Changes(self):
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.createPrimWithAttribute('/A')
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/A']])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.changePrimAttribute('/A', 2)
        self.assertItemsEqual(self._changed, [Sdf.Path(x) for x in ['/A.prop']])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.changePrimAttribute('/A', 4)
            self.changePrimAttribute('/A', 2) ## effectively no change
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [])

    ## Test that CloseNotice reports hierarchy changes as expected
    def test_Hierarchy(self):
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.createPrimWithAttribute('/root')
            self.createPrimWithAttribute('/root/A')
            self.createPrimWithAttribute('/root/A/C')
            self.createPrimWithAttribute('/root/A/D')
            self.createPrimWithAttribute('/root/B')
            self.createPrimWithAttribute('/root/B/E')
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/root']])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.changePrimAttribute('/root', 2)
            self.changePrimAttribute('/root/A/D', 2)
        self.assertItemsEqual(self._changed, [Sdf.Path(x) for x in ['/root.prop', '/root/A/D.prop']])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.changePrimAttribute('/root/B', 2)
            self.changePrimAttribute('/root/A/C', 2)
            self.changePrimAttribute('/root/A/C', 1) ## effectively no change
        self.assertItemsEqual(self._changed, [Sdf.Path(x) for x in ['/root/B.prop']])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.createPrimWithAttribute('/root/B/F')
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/root/B/F']])

    ## Test that CloseNotice reports clearing layers as expected
    def test_Clear(self):
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.createPrimWithAttribute('/root')
            self.createPrimWithAttribute('/root/A')
            self.createPrimWithAttribute('/root/A/B')
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/root']])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self._stage.GetSessionLayer().Clear()
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/root']])
    
        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self.createPrimWithAttribute('/root')
            self.createPrimWithAttribute('/root/A')
            self.createPrimWithAttribute('/root/A/B')
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [Sdf.Path(x) for x in ['/root']])

        with transaction.ScopedTransaction(self._stage, self._stage.GetSessionLayer()):
            self._stage.GetSessionLayer().Clear()
            self.createPrimWithAttribute('/root')
            self.createPrimWithAttribute('/root/A')
            self.createPrimWithAttribute('/root/A/B')
            ### effectively no change
        self.assertItemsEqual(self._changed, [])
        self.assertItemsEqual(self._resynced, [])


if __name__ == '__main__':
    unittest.main()
